group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/ServerSideStickyRangeAssignor.java

```diff
@@ -0,0 +1,268 @@
+ * <li> 5) Iterate through unfilled consumers and assign partitions from available partitions </li>
+ * </ul>
+ * </p>
+ *
+ */
+package org.apache.kafka.coordinator.group.assignor;
+
+
+import org.apache.kafka.common.Uuid;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.HashMap;
+import java.util.Collections;
+
+import static java.lang.Math.min;
+
+public class ServerSideStickyRangeAssignor implements PartitionAssignor {
+
+    public static final String RANGE_ASSIGNOR_NAME = "range-sticky";
+
+    @Override
+    public String name() {
+        return RANGE_ASSIGNOR_NAME;
+    }
+
+    protected static <K, V> void putList(Map<K, List<V>> map, K key, V value) {
+        List<V> list = map.computeIfAbsent(key, k -> new ArrayList<>());
+        list.add(value);
+    }
+
+    protected static <K, V> void putSet(Map<K, Set<V>> map, K key, V value) {
+        Set<V> set = map.computeIfAbsent(key, k -> new HashSet<>());
+        set.add(value);
+    }
+
+    static class Pair<T, U> {
+        private final T first;
+        private final U second;
+
+        public Pair(T first, U second) {
+            this.first = first;
+            this.second = second;
+        }
+
+        public T getFirst() {
+            return first;
+        }
+
+        public U getSecond() {
+            return second;
+        }
+
+        @Override
+        public String toString() {
+            return "(" + first + ", " + second + ")";
+        }
+    }
+
+    // Returns a map of the list of consumers per Topic (keyed by topicId)
+    private Map<Uuid, List<String>> consumersPerTopic(AssignmentSpec assignmentSpec) {
+        Map<Uuid, List<String>> mapTopicsToConsumers = new HashMap<>();
+        Map<String, AssignmentMemberSpec> membersData = assignmentSpec.members;
+
+        for (Map.Entry<String, AssignmentMemberSpec> memberEntry : membersData.entrySet()) {
+            String memberId = memberEntry.getKey();
+            AssignmentMemberSpec memberMetadata = memberEntry.getValue();
+            List<Uuid> topics = new ArrayList<>(memberMetadata.subscribedTopics);
+            for (Uuid topicId: topics) {
+                putList(mapTopicsToConsumers, topicId, memberId);
+            }
+        }
+        return mapTopicsToConsumers;
+    }
+
+    private Map<Uuid, List<Integer>> getAvailablePartitionsPerTopic(AssignmentSpec assignmentSpec, Map<Uuid, Set<Integer>> assignedStickyPartitionsPerTopic) {
+        Map<Uuid, List<Integer>> availablePartitionsPerTopic = new HashMap<>();
+        Map<Uuid, AssignmentTopicMetadata> topicsMetadata = assignmentSpec.topics;
+        // Iterate through the topics map provided in assignmentSpec
+        for (Map.Entry<Uuid, AssignmentTopicMetadata> topicMetadataEntry : topicsMetadata.entrySet()) {
+            Uuid topicId = topicMetadataEntry.getKey();
+            availablePartitionsPerTopic.put(topicId, new ArrayList<>());
```

Comment:
can we create a new arraylist and add it at the end instead of accessing the map each time to add a new partition in L131?

Reply:
changed it!